From d7d5f988f5c57011a4fe6ce349430ba417a864b5 Mon Sep 17 00:00:00 2001
From: Eskil Abrahamsen Blomfeldt <eskil.abrahamsen-blomfeldt@qt.io>
Date: Thu, 22 Apr 2021 08:42:33 +0200
Subject: [PATCH 06/20] client: Gracefully handle shutdown and window hiding

When a window is hidden or destroyed, the render thread may already
be rendering. We need to properly read-lock the surface pointer
when it is in use and exit when it becomes null.

Note that there is also a potential crash in the Mesa GL driver
where it keeps a proxy to the wl_surface, so if we delete this
while we are still rendering, it can crash inside the driver.
This is not addressed by this patch, and has not been reproduced
on any other drivers so far.

[ChangeLog][Client] Fixed a crash that could happen when hiding
or closing windows while Qt Quick was actively rendering on
a different thread.

Pick-to: 6.0 6.1 5.15
Fixes: QTBUG-91264
Fixes: QTBUG-90037
Task-number: QTBUG-92249
Change-Id: I029b123b83c58740321e8b90a463ced748d8bcf4
Reviewed-by: Laszlo Agocs <laszlo.agocs@qt.io>
Reviewed-by: David Edmundson <davidedmundson@kde.org>

Conflicts:
	src/client/qwaylandwindow.cpp
	src/hardwareintegration/client/wayland-egl/qwaylandglcontext.cpp

(cherry picked from commit b19b0fbaf775e8b8eda1e03c265a5393d618c6c0)
Signed-off-by: Jeffy Chen <jeffy.chen@rock-chips.com>
---
 src/client/qwaylandwindow.cpp | 13 ++++++++++++-
 src/client/qwaylandwindow_p.h |  2 +-
 2 files changed, 13 insertions(+), 2 deletions(-)

diff --git a/src/client/qwaylandwindow.cpp b/src/client/qwaylandwindow.cpp
index c9d4a873..d548fbe0 100644
--- a/src/client/qwaylandwindow.cpp
+++ b/src/client/qwaylandwindow.cpp
@@ -371,6 +371,10 @@ void QWaylandWindow::closePopups(QWaylandWindow *parent)
 
 QWaylandScreen *QWaylandWindow::calculateScreenFromSurfaceEvents() const
 {
+    QReadLocker lock(&mSurfaceLock);
+    if (!isInitialized())
+        return screen();
+
     return mScreens.isEmpty() ? screen() : mScreens.first();
 }
 
@@ -420,6 +424,7 @@ void QWaylandWindow::setMask(const QRegion &mask)
 
     mMask = mask;
 
+    QReadLocker locker(&mSurfaceLock);
     if (!isInitialized())
         return;
 
@@ -591,6 +596,10 @@ void QWaylandWindow::attachOffset(QWaylandBuffer *buffer)
 
 void QWaylandWindow::damage(const QRect &rect)
 {
+    QReadLocker locker(&mSurfaceLock);
+    if (!isInitialized())
+        return;
+
     damage(rect.x(), rect.y(), rect.width(), rect.height());
 }
 
@@ -616,6 +625,7 @@ void QWaylandWindow::handleExpose(const QRegion &region)
 
 void QWaylandWindow::commit(QWaylandBuffer *buffer, const QRegion &damage)
 {
+    QReadLocker locker(&mSurfaceLock);
     if (!isInitialized())
         return;
 
@@ -680,7 +690,8 @@ QWaylandScreen *QWaylandWindow::screen() const
 
 void QWaylandWindow::handleContentOrientationChange(Qt::ScreenOrientation orientation)
 {
-    if (mDisplay->compositorVersion() < 2)
+    QReadLocker locker(&mSurfaceLock);
+    if (!isInitialized() || mDisplay->compositorVersion() < 2)
         return;
 
     wl_output_transform transform;
diff --git a/src/client/qwaylandwindow_p.h b/src/client/qwaylandwindow_p.h
index 41f8f185..8d412b04 100644
--- a/src/client/qwaylandwindow_p.h
+++ b/src/client/qwaylandwindow_p.h
@@ -278,7 +278,7 @@ private:
     static QMutex mFrameSyncMutex;
     static QWaylandWindow *mMouseGrab;
 
-    QReadWriteLock mSurfaceLock;
+    mutable QReadWriteLock mSurfaceLock;
 
     friend class QWaylandSubSurface;
 };
-- 
2.20.1

